src/js/autocomplete.js

```diff
@@ -1,14 +1,247 @@
+ * @returns {void}
+ */
+
+/**
+ * @callback Suggestions
+ * @param {string} query - Text which was typed into the autocomplete by the user
+ * @param {PopulateResults} populateResults - Function to call when ready to update the suggestions dropdown
+ * @returns {void}
+*/
+
+
+/**
+ * @typedef {Object} AutocompleteOptions
+ * @property {Suggestions} source - The function which retrieves the suggestions to display
+ */
+
 class Autocomplete {
 	/**
 	 * Class constructor.
 	 * @param {HTMLElement} [autocompleteEl] - The component element in the DOM
-	 * @param {Object} [options={}] - An options object for configuring the component
+	 * @param {AutocompleteOptions} [options={}] - An options object for configuring the component
 	 */
 	constructor (autocompleteEl, options) {
 		this.autocompleteEl = autocompleteEl;
-		this.options = Object.assign({}, {
+		this.options = Object.assign({
+			placeholder: '',
+			cssNamespace: 'o-autocomplete',
+			displayMenu: 'overlay',
+			showNoOptionsFound: false,
+			templates: {
+				suggestion: this.suggestionTemplate.bind(this)
+			}
 		}, options || Autocomplete.getDataAttributes(autocompleteEl));
+
+		if (this.options.source) {
+			// If source is a string, then it is the name of a global function to use.
+			// If source is not a string, then it is a function to use.
+			/**
+			 * @type {Suggestions}
+			 */
+			const customSuggestions = typeof this.options.source === 'string' ? window[this.options.source] : this.options.source;
+			/**
+			 * @param {string} query - Text which was typed into the autocomplete by the user
+			 * @param {PopulateResults} populateResults - Function to call when ready to update the suggestions dropdown
+			 * @returns {void}
+			*/
+			this.options.source = (query, populateResults) => {
+				this.showLoadingPane();
+				/**
+				 * @param {Array<string>} results - The results to show in the suggestions dropdown
+				 * @returns {void}
+				 */
+				const callback = (results) => {
+					this.hideLoadingPane();
+					populateResults(results);
+				};
+				customSuggestions(query, callback);
+			};
+			const input = autocompleteEl.querySelector('input');
+			const id = input.getAttribute('id');
+			autocompleteEl.innerHTML = '';
+			const options = Object.assign({
+				element: autocompleteEl,
+				id: id,
+			}, this.options);
+			accessibleAutocomplete(options);
+		} else {
+			const element = autocompleteEl.querySelector('select');
+			autocompleteEl.appendChild(element);
+			const options = Object.assign({
+				selectElement: element,
+				defaultValue: '',
+			}, this.options);
+			options.autoselect = false; // TODO: Find out if we should allow autoselect/hinting in the input
+			accessibleAutocomplete.enhanceSelectElement(options);
+			element.parentElement.removeChild(element); // Remove the original select element
+		}
+
+		this.loadingContainer = createLoadingContainer();
+		this.initClearButton();
+	}
+
+	/**
+	 * Show the loading panel
+	 * @private
```

Comment:
Non-blocking suggestion/question: I was about to comment that these should be private, but see you handled that with JSDoc :) I worry a bit that projects may come to rely on them accidentality if users don't consult the docs/have hinting. My preference would be to additionally indicate they are private by prepending an underscore to their name, as some other components do, or take them off the prototype and pass in the autocomplete instance to make them actually private

I think it's fine as is though, I wouldn't feel toooo guilty telling any projects that break due to relying on methods documented as private to not do that, what are your thoughts?

Reply:
Happy to have them pulled out of the prototype and become truly private functions which take the instance as an argument. Talking about private methods -- Private class methods are going to be added to the ecma-2022 specification, they are already at stage 4 of the process -- https://github.com/tc39/proposals/blob/master/finished-proposals.md#finished-proposals

Reply:
Done in [`a97cf23` (#2)](https://github.com/Financial-Times/o-autocomplete/pull/2/commits/a97cf2316cb1e52ae17314c09641196918ac7187)